<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [124]:
import pandas as pd
import psycopg2

In [125]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [181]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [127]:
# текст запроса
query_3_1 = f'''
select *
from 
    public.vacancies
'''

In [128]:
# результат запроса
df_vacan = pd.read_sql_query(query_3_1, connection)
vac_cou = df_vacan['id'].count()
print(f'Количество вакансий в базе - {vac_cou}')

c:\Users\cch-o\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


Количество вакансий в базе - 49197


2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [129]:
# текст запроса
query_3_2 = f'''
select *
from 
    public.employers
'''

In [130]:
# результат запроса
df_empi = pd.read_sql_query(query_3_2, connection)
emp_cou = df_empi['id'].count()
print(f'Количество работадателей в базе - {emp_cou}')

c:\Users\cch-o\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


Количество работадателей в базе - 23501


3. Посчитате с помощью запроса количество регионов (таблица areas).

In [131]:
# текст запроса
query_3_3 = f'''
select *
from 
    public.areas
'''

In [132]:
# результат запроса
df_areas = pd.read_sql_query(query_3_3, connection)
are_cou = df_areas['id'].count()
print(f'Количество регионов в базе - {are_cou}')

Количество регионов в базе - 1362


c:\Users\cch-o\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


4. Посчитате с помощью запроса количество сфер деятельности (таблица industries).

In [133]:
# текст запроса
query_3_4 = f'''
select *
from 
    public.industries
'''

In [134]:
# результат запроса
df_ind = pd.read_sql_query(query_3_4, connection)
ind_cou = df_ind['id'].count()
print(f'Количество сфер деятельности в базе - {ind_cou}')

Количество сфер деятельности в базе - 294


c:\Users\cch-o\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


***

### Выводы: Из полученных результатом мы видим размерность таблиц базы данных. Таблица "вакансии" имеет размерность 10 столбцов и 49197 строк; в столбцах salary_to, salary_from, key_skills - есть пропуски. Таблица "Работадатели" имеет размерность 3 столбца и 23501 строк, без пропусков. Таблица "Регионы" имеет размерность 2 столбца и 1362 строк, без пропусков. Таблица "Сферы деятельности" имеет размерность 2 столбца и 294 строки, без пропусков.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [182]:
# текст запроса
query_4_1 = f'''
select
a.name,
count(v.name)
from 
     public.areas a
join public.vacancies v on a.id = v.area_id
Group BY a.name
Order by count(v.name) desc
'''

In [185]:
# результат запроса
df_4_1 = pd.read_sql_query(query_4_1, connection)
df_4_1.head(5)

c:\Users\cch-o\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name  count
0           Москва   5333
1  Санкт-Петербург   2851
2            Минск   2112
3      Новосибирск   2006
4           Алматы   1892

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [137]:
# текст запроса
query_4_2 = f'''
select
count(v.name)
from 
     public.vacancies v
where salary_from is not null or salary_to is not null
'''

In [138]:
# результат запроса
df_4_2 = pd.read_sql_query(query_4_2, connection)
df_4_2_1 = df_4_2['count'].values [0]
print(f'Количество вакансий у которых заполнено хотя бы одно из двух полей с зарплатой равна -  {round(df_4_2_1)}.')

Количество вакансий у которых заполнено хотя бы одно из двух полей с зарплатой равна -  24073.


c:\Users\cch-o\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [139]:
# текст запроса
query_4_3 = f'''
select
AVG(v.salary_from) mean_salary_from,
AVG(v.salary_to) mean_salary_to
from 
     public.vacancies v
'''

In [140]:
# результат запроса
df_4_3 = pd.read_sql_query(query_4_3, connection)
df_4_3_1 = df_4_3['mean_salary_from'].values [0]
df_4_3_2 = df_4_3['mean_salary_to'].values [0]
print(f'Среднее значение для нижней границы зарплатной вилки равно -  {round(df_4_3_1)} рублей, а для верхней границы - {round(df_4_3_2)} рублей.')


Среднее значение для нижней границы зарплатной вилки равно -  71065 рублей, а для верхней границы - 110537 рублей.


c:\Users\cch-o\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [186]:
# текст запроса
query_4_4 = f'''
select
concat(v.schedule, ' - ', v.employment) new_col,
count(v.name) count_col
from 
     public.vacancies v
group by new_col
order by count_col desc
'''

In [187]:
# результат запроса
df_4_4 = pd.read_sql_query(query_4_4, connection)
df_4_4

c:\Users\cch-o\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


new_col  count_col
0           Полный день - Полная занятость      35367
1      Удаленная работа - Полная занятость       7802
2         Гибкий график - Полная занятость       1593
3   Удаленная работа - Частичная занятость       1312
4        Сменный график - Полная занятость        940
5                 Полный день - Стажировка        569
6        Вахтовый метод - Полная занятость        367
7        Полный день - Частичная занятость        347
8      Гибкий график - Частичная занятость        312
9           Полный день - Проектная работа        141
10     Удаленная работа - Проектная работа        133
11              Гибкий график - Стажировка        116
12    Сменный график - Частичная занятость        101
13           Удаленная работа - Стажировка         64
14        Гибкий график - Проектная работа         18
15             Сменный график - Стажировка         12
16       Вахтовый метод - Проектная работа          2
17       Сменный график - Проектная работа          1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [143]:
# текст запроса
query_4_5 = f'''
select
v.experience,
count(v.name) coun_name
from 
     public.vacancies v
group by v.experience
order by coun_name
'''

In [144]:
# результат запроса
df_4_5 = pd.read_sql_query(query_4_5, connection)
df_4_5

c:\Users\cch-o\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


experience  coun_name
0         Более 6 лет       1337
1           Нет опыта       7197
2       От 3 до 6 лет      14511
3  От 1 года до 3 лет      26152

***

### Вывод: Проведя анализ вакансий можно сделать выводы: самое большое количество вакансий находится в Москве и С-Петербурге. Средняя зарплата находится в промежутке от 71000 до 110000, требуемый опыт работы от 1 до 3 лет. На ряду с полным рабочим днем много вакансий на удале5нную работу.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [145]:
# текст запроса
query_5_1 = f'''
select
e.name,
count(v.name)
from 
     public.employers e
join public.vacancies v on e.id = v.employer_id
Group BY e.name
Order by count(v.name) desc
'''

In [146]:
# результат запроса
df_5_1 = pd.read_sql_query(query_5_1, connection)
df_5_1_1 = df_5_1['name'].values [0]
df_5_1_2 = df_5_1['name'].values [4]

print(f'На первом месте по количеству вакансий находится "{df_5_1_1}", а на пятом месте - "{df_5_1_2}".')

c:\Users\cch-o\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


На первом месте по количеству вакансий находится "Яндекс", а на пятом месте - "Газпром нефть".


2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [147]:
# текст запроса
query_5_2 = f'''
select
a.name region,
count(distinct e.id) employer, 
count(distinct v.id) vacanci
from public.areas a 
	left join public.employers e on e.area = a.id
    left join public.vacancies v on v.area_id=a.id
where v.id is null
group by region 
having count(distinct v.id)=0
order by vacanci asc, employer desc
'''

In [148]:
# результат запроса
df_5_2 = pd.read_sql_query(query_5_2, connection)
df_5_2_1 = df_5_2['region'].values [0]
df_5_2_2 = df_5_2['employer'].values [0]
print(f'Самый большой регион (поданным запроса) в котором нет ввакансий носамое большое колличество работадателей - {df_5_2_1}. В данном регионе {df_5_2_2} работадателей.')

Самый большой регион (поданным запроса) в котором нет ввакансий носамое большое колличество работадателей - Россия. В данном регионе 410 работадателей.


c:\Users\cch-o\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [149]:
# текст запроса
query_5_3 = f'''
select
e.name employers_name,
count(distinct v.area_id) area_count
from public.vacancies v
    join public.employers e on e.id = v.employer_id
group by employers_name 
order by area_count desc
'''

In [189]:
# результат запроса
df_5_3 = pd.read_sql_query(query_5_3, connection)
df_5_3.head(5)

c:\Users\cch-o\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


employers_name  area_count
0                  Яндекс         181
1              Ростелеком         152
2              Спецремонт         116
3  Поляков Денис Иванович          88
4               ООО ЕФИН           71

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [151]:
# текст запроса
query_5_4 = f'''
SELECT
count(e.name) count_empploer
FROM
    public.EMPLOYERS e
    LEFT JOIN public.EMPLOYERS_INDUSTRIES e_i on e_i.employer_id = e.id
where e_i.industry_id is NULL
'''


In [196]:
# результат запроса
df_5_4 = pd.read_sql_query(query_5_4, connection)
df_5_4_1 = df_5_4['count_empploer'].values [0]
print(f'Работадателей, у которых не указана сфера деятельности - {df_5_4_1}.')

Работадателей, у которых не указана сфера деятельности - 8419.


c:\Users\cch-o\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [153]:
# текст запроса
query_5_5 = f'''
SELECT
e.name employer_name,
count(i.id) count_ind
FROM
    public.EMPLOYERS e
    LEFT JOIN public.EMPLOYERS_INDUSTRIES e_i on e_i.employer_id = e.id
    LEFT JOIN public.INDUSTRIES i on e_i.industry_id = i.id
group by employer_name
having count(i.id) = 4
order by employer_name
OFFSET 2 limit 1
'''

In [199]:
# результат запроса
df_5_5 = pd.read_sql_query(query_5_5, connection)
df_5_5_1 = df_5_5['employer_name'].values [0]
print(f'Компаниия находящеяся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности - {df_5_5_1}.')

Компаниия находящеяся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности - 2ГИС.


c:\Users\cch-o\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [155]:
# текст запроса
query_5_6 = f'''
SELECT
i.name,
count(e.id)
FROM
    public.EMPLOYERS e
    LEFT JOIN public.EMPLOYERS_INDUSTRIES e_i on e_i.employer_id = e.id
    LEFT JOIN public.INDUSTRIES i on e_i.industry_id = i.id
where i.name = 'Разработка программного обеспечения' 
group by i.name
'''

In [202]:
# результат запроса
df_5_6 = pd.read_sql_query(query_5_6, connection)
df_5_6_1 = df_5_6['count'].values [0]
print(f'У {df_5_6_1} работодателей в качестве сферы деятельности указана "Разработка программного обеспечения".')

У 3553 работодателей в качестве сферы деятельности указана "Разработка программного обеспечения".


c:\Users\cch-o\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [157]:
import requests 
from bs4 import BeautifulSoup

In [158]:
url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8'
response = requests.get(url) # Выполняем GET-запрос, содержимое ответа присваивается переменной response
page = BeautifulSoup(response.text, 'html.parser') # Создаём объект BeautifulSoup, указывая html-парсер
list_city = page.find_all('table') # находим все таблицы на странице
list_city = list_city[1] # выделяем интересующую нас таблицу городов-миллиоников
lst_city = [] # создаем список для занесения данных
for i in list_city.find_all('td'): # перепираем все вхождения 'td' в таблице
    title = i.text  # значением переводим в текст
    lst_city.append(title) # полученные результаты заносим в список
lst_city = lst_city[1::5] # в полученном списке выделяем только наши города без других данных таблицы
lst_city[13] = 'Воронеж' # заменим Воронеж[а] на Воронеж
lst_city = tuple(lst_city)
print(lst_city)

('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Челябинск', 'Красноярск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Омск', 'Краснодар', 'Воронеж', 'Пермь', 'Волгоград')


***

In [159]:
query_5_7 = f'''
(select 
    a.name "Город",
    count(v.id) 
from vacancies v
    join employers e on e.id = v.employer_id
    join areas a on a.id = v.area_id
where
    e.name = 'Яндекс'
and a.name in  {str(lst_city)}
group by 1
)

union 

(select 
    'Total',
    count(v.id)
from vacancies v
    join employers e on e.id = v.employer_id
    join areas a on a.id = v.area_id
where
    e.name = 'Яндекс'
and a.name in {str(lst_city)}
)
order by 2 
'''

In [218]:
df_5_7 = pd.read_sql_query(query_5_7, connection)
df_5_7.head(20)

c:\Users\cch-o\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


Город  count
0              Омск     21
1         Челябинск     22
2        Красноярск     23
3         Волгоград     24
4             Пермь     25
5            Казань     25
6    Ростов-на-Дону     25
7               Уфа     26
8            Самара     26
9         Краснодар     30
10          Воронеж     32
11      Новосибирск     35
12  Нижний Новгород     36
13     Екатеринбург     39
14  Санкт-Петербург     42
15           Москва     54
16            Total    485

### Вывод: Проведя анализ делаем вывод, что на первом месте по количеству вакансий и количеству регионов где представлены вакансии - Яндекс.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [162]:
# текст запроса
query_6_1 = f'''
select
count(v.name)
from 
     public.VACANCIES v
where lower(v.name) LIKE '%data%' OR lower(v.name) LIKE '%данн%'
'''

In [206]:
# результат запроса
df_6_1 = pd.read_sql_query(query_6_1, connection)
df_6_1_1 = df_6_1['count'].values [0]
print(f'{df_6_1_1} вакансия имеет отношение к данным.')

1771 вакансия имеет отношение к данным.


c:\Users\cch-o\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [164]:
# текст запроса
query_6_2 = f'''
select
    count(v.name)
from 
    vacancies v
where (lower(v.name) like '%data%scientist%'
    or lower(v.name) like '%data%science%'
    or lower(v.name) like '%исследователь%данных%'
    or (v.name like '%ML%'
    and v.name not like '%HTML%')
    or lower(v.name) like '%machine%learning%'
    or lower(v.name) like '%машинн%обучен%')
    and (lower(v.experience) like '%нет опыта%' 
    or lower(v.employment) like '%стажировка%'
    or lower(v.name) like '%junior%')
'''

In [211]:
# результат запроса
df_6_2 = pd.read_sql_query(query_6_2, connection)
df_6_2_1 = df_6_2['count'].values [0]
print(f'В базе работодателей расположена {df_6_2_1} вакансия для начинающего дата-сайентиста.')

В базе работодателей расположена 51 вакансия для начинающего дата-сайентиста.


c:\Users\cch-o\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [166]:
# текст запроса
query_6_3 = f'''
select
    count(v.name)
from 
    vacancies v
where (lower(v.name) like '%data%scientist%'
    or lower(v.name) like '%data%science%'
    or lower(v.name) like '%исследователь%данных%'
    or (v.name like '%ML%'
    and v.name not like '%HTML%')
    or lower(v.name) like '%machine%learning%'
    or lower(v.name) like '%машинн%обучен%')
    and (v.key_skills like '%SQL%' 
    or lower(v.key_skills) like '%postgres%')
'''

In [212]:
# результат запроса\
df_6_3 = pd.read_sql_query(query_6_3, connection)
df_6_3_1 = df_6_3['count'].values [0]
print(f'В базе работодателей расположена {df_6_3_1} вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres.')

В базе работодателей расположена 201 вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres.


c:\Users\cch-o\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [215]:
# текст запроса
query_6_4 = f'''
select
    count(v.name) количество
from 
    vacancies v
where (lower(v.name) like '%data%scientist%'
    or lower(v.name) like '%data%science%'
    or lower(v.name) like '%исследователь%данных%'
    or (v.name like '%ML%'
    and v.name not like '%HTML%')
    or lower(v.name) like '%machine%learning%'
    or lower(v.name) like '%машинн%обучен%')
    and 
    lower(v.key_skills) like '%python%'
'''

In [216]:
# результат запроса
df_6_4 = pd.read_sql_query(query_6_4, connection)
df_6_4_1 = df_6_4['количество'].values [0]
print(f'В базе работодателей расположена {df_6_4_1} вакансий, в которых в качестве ключевого навыка указан Python.')


В базе работодателей расположена 351 вакансий, в которых в качестве ключевого навыка указан Python.


c:\Users\cch-o\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [170]:
# текст запроса
query_6_5 = f'''
SELECT
    AVG(LENGTH(v.key_skills) - LENGTH(replace(v.key_skills,CHR(9), ''))+1) 
    FROM
    public.VACANCIES v /*из таблицы вакансий*/
    where (lower(v.name) LIKE '%data scientist%'
    OR lower (v.name) LIKE '%data science%'
    or lower (v.name) LIKE '%исследователь данных%'
    or (v.name LIKE '%ML%' and v.name not LIKE '%HTML%')
    or lower (v.name) LIKE '%machine learning%'
    or lower (v.name) LIKE '%машинн%обучен%%')
    and not v.key_skills is NULL
'''

In [217]:
# результат запроса
df_6_5 = pd.read_sql_query(query_6_5, connection)
df_6_5_1 = df_6_5['avg'].values [0]
print(f'{round(df_6_5_1,2)} ключевых навыков в среднем указывают в вакансиях для DS.')

6.41 ключевых навыков в среднем указывают в вакансиях для DS.


c:\Users\cch-o\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [172]:
# текст запроса
query_6_6 = f'''
SELECT
    experience,
    AVG(COALESCE((salary_from+salary_to)/2, salary_from, salary_to,0))
FROM public.vacancies v
WHERE (lower(name) LIKE '%data scientist%'
    OR lower(name) LIKE '%data science%'
    OR lower(name) LIKE '%исследователь данных%'
    OR (name LIKE '%ML%' AND name NOT LIKE '%HTML%')
    OR lower(name) LIKE '%machine learning%'
    OR lower(name) LIKE '%машинн%обучен%%')
    AND (salary_from IS NOT NULL OR salary_to IS NOT NULL)
    AND experience like 'От 3 до 6 лет'
GROUP BY experience
'''

In [173]:
# результат запроса
df_6_6 = pd.read_sql_query(query_6_6, connection)
df_6_6_1 = df_6_6['avg'].values [0]
print(f'Средняя зарплата дата-сайентиста с опытом работы от 3 до 6 лет равна -  {round(df_6_6_1)} рублей.')

Средняя зарплата дата-сайентиста с опытом работы от 3 до 6 лет равна -  243115 рублей.


c:\Users\cch-o\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\io\sql.py:761: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


In [174]:
emp_cou111 = df_6_6['avg'].values [0]
print(f'Количество работадателей в базе - {round(emp_cou111)}')

Количество работадателей в базе - 243115


***

### Вывод: Проведя предметный анализ рынка труда на предмет работы с данными получили следующее: 1771 вакансий так или иначе связана с данными, что в свою очередь является не большим количеством по сравнению с общим количеством вакансий и является новым и развивающимся направлением. ДЛя начинающего DS выявлено всего 51 вакансия. В среднем специалист по DS должен обладать 6 ключевыми навыками. Так для примера 201 вакансия где ключивыми навыками является SQL и 351 Python. Средняя заработная плата DS с опытом работы 3-6 лет - 243 тысячи р.

In [176]:
connection.close()

# Общий вывод по проекту

### Самое большое количество вакансий находится в Москве и С-Петербурге. Средняя зарплата находится в промежутке от 71000 до 110000, требуемый опыт работы от 1 до 3 лет. На ряду с полным рабочим днем много вакансий на удале5нную работу. Из работадателей первом месте по количеству вакансий и количеству регионов где представлены вакансии - является Яндекс. Не большое колличество вакансий связанно с данными. Для начинающего DS вакансий не много. 